nexus/tests/integration_tests/volume_management.rs

```diff
@@ -1813,6 +1813,132 @@ async fn test_volume_checkout_updates_sparse_mid_multiple_gen(
     volume_match_gen(new_vol, vec![Some(8), None, Some(10)]);
 }
 
+#[nexus_test]
+async fn test_disk_create_saga_unwinds_correctly(
+    cptestctx: &ControlPlaneTestContext,
+) {
+    // Test that Nexus properly unwinds when not all regions are successfully
+    // created.
+
+    let client = &cptestctx.external_client;
+    populate_ip_pool(&client, "default", None).await;
+    create_org_and_project(client).await;
+
+    let disk_test = DiskTest::new(&cptestctx).await;
+    let disks_url = get_disks_url();
+    let base_disk_name: Name = "base-disk".parse().unwrap();
+
+    // Set the third agent to fail creating the region
+    let zpool = &disk_test.zpools[2];
+    let dataset = &zpool.datasets[0];
+    disk_test
+        .sled_agent
+        .get_crucible_dataset(zpool.id, dataset.id)
+        .await
+        .set_region_creation_limit(0)
+        .await;
+
+    let disk_size = ByteCount::from_gibibytes_u32(2);
+    let base_disk = params::DiskCreate {
+        identity: IdentityMetadataCreateParams {
+            name: base_disk_name.clone(),
+            description: String::from("sells rainsticks"),
+        },
+        disk_source: params::DiskSource::Blank {
+            block_size: params::BlockSize::try_from(512).unwrap(),
+        },
+        size: disk_size,
+    };
+
+    NexusRequest::new(
+        RequestBuilder::new(client, Method::POST, &disks_url)
+            .body(Some(&base_disk))
+            .expect_status(Some(StatusCode::INTERNAL_SERVER_ERROR)),
+    )
+    .authn_as(AuthnMode::PrivilegedUser)
+    .execute()
+    .await
+    .unwrap()
+    .parsed_body::<dropshot::HttpErrorResponseBody>()
+    .unwrap();
+
+    // Assert everything was cleaned up
+    assert!(disk_test.crucible_resources_deleted().await);
+}
+
+#[nexus_test]
+async fn test_snapshot_create_saga_unwinds_correctly(
+    cptestctx: &ControlPlaneTestContext,
+) {
+    // Test that Nexus properly unwinds when not all regions are successfully
+    // created.
+
+    let client = &cptestctx.external_client;
+    populate_ip_pool(&client, "default", None).await;
+    create_org_and_project(client).await;
+
+    let disk_test = DiskTest::new(&cptestctx).await;
+    let disks_url = get_disks_url();
+    let base_disk_name: Name = "base-disk".parse().unwrap();
+
+    // Set the third agent to fail creating the region
+    let zpool = &disk_test.zpools[2];
+    let dataset = &zpool.datasets[0];
+    disk_test
+        .sled_agent
+        .get_crucible_dataset(zpool.id, dataset.id)
+        .await
+        .set_region_creation_limit(1)
+        .await;
```

Comment:
I assume this is set to 1 so that the disk creation will succeed while the snapshot request fails. The intent might be a little clearer if you just set it to 0 after creating the disk. (At that point you might not even want this to be a numerical limit: instead it could just be a `bool`, or, if you really want to gild the lily, you could tell the simulated sled agent exactly what kind of error to return instead of having it return 500 Internal Server Error unconditionally.)

Reply:
Good idea - I like the bool better, as the simulated sled agent is only returning 500s at that point. Changed in cc260f2f5